sqlglot/optimizer/pushdown_predicates.py

```diff
@@ -18,58 +20,116 @@ def pushdown_predicates(expression):
     Returns:
         sqlglot.Expression: optimized expression
     """
+    expression = simplify(expression)
+
     for scope in reversed(traverse_scope(expression)):
         where = scope.expression.args.get("where")
 
         if not where:
             continue
 
         condition = where.this.unnest()
-        predicates = (
-            condition.flatten() if isinstance(condition, exp.And) else [condition]
-        )
 
-        for predicate in predicates:
-            sources = [
-                scope.sources.get(table) for table in exp.column_table_names(predicate)
-            ]
+        cnf_like = normalized(condition) or not normalized(condition, dnf=True)
+
+        predicates = list(
+            condition.flatten()
+            if isinstance(condition, exp.And if cnf_like else exp.Or)
+            else [condition]
+        )
 
-            if len(sources) != 1:
-                continue
+        if cnf_like:
+            for predicate in predicates:
+                for node in nodes_for_predicate(predicate, scope).values():
+                    if isinstance(node, exp.Join):
+                        predicate.replace(exp.TRUE)
+                        node.on(predicate, copy=False)
+                        break
+                    if isinstance(node, exp.Select):
+                        predicate.replace(exp.TRUE)
+                        node.where(replace_aliases(node, predicate), copy=False)
+        else:
+            pushdown = set()
```

Comment:
This block is very hard to follow. 

I know complexity is complex... but maybe we can throw some comments here and there?